cmd/kubeadm/app/cmd/init.go

```diff
@@ -106,48 +107,71 @@ var (
+	cfgPath               string
+	skipTokenPrint        bool
+	dryRun                bool
+	featureGatesString    string
+	ignorePreflightErrors []string
+	bto                   *options.BootstrapTokenOptions
+	externalcfg           *kubeadmapiv1beta1.InitConfiguration
+}
+
+// initData defines all the runtime information used when running the kubeadm init worklow;
+// this data is shared across all the phases that are included in the workflow.
+type initData struct {
+	cfg                   *kubeadmapi.InitConfiguration
+	skipTokenPrint        bool
+	dryRun                bool
+	ignorePreflightErrors sets.String
+}
+
 // NewCmdInit returns "kubeadm init" command.
 func NewCmdInit(out io.Writer) *cobra.Command {
-	externalcfg := &kubeadmapiv1beta1.InitConfiguration{}
-	kubeadmscheme.Scheme.Default(externalcfg)
-
-	var cfgPath string
-	var skipTokenPrint bool
-	var dryRun bool
-	var featureGatesString string
-	var ignorePreflightErrors []string
-	// Create the options object for the bootstrap token-related flags, and override the default value for .Description
-	bto := options.NewBootstrapTokenOptions()
-	bto.Description = "The default bootstrap token generated by 'kubeadm init'."
+	options := newInitOptions()
+	initRunner := workflow.NewRunner()
 
 	cmd := &cobra.Command{
 		Use:   "init",
 		Short: "Run this command in order to set up the Kubernetes master.",
 		Run: func(cmd *cobra.Command, args []string) {
-			var err error
-			if externalcfg.FeatureGates, err = features.NewFeatureGate(&features.InitFeatureGates, featureGatesString); err != nil {
-				kubeadmutil.CheckErr(err)
-			}
-
-			ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(ignorePreflightErrors)
+			c, err := initRunner.InitData()
 			kubeadmutil.CheckErr(err)
 
-			err = validation.ValidateMixedArguments(cmd.Flags())
-			kubeadmutil.CheckErr(err)
+			ctx, _ := c.(initData)
+			fmt.Printf("[init] using Kubernetes version: %s\n", ctx.cfg.KubernetesVersion)
 
-			err = bto.ApplyTo(externalcfg)
+			err = initRunner.Run()
 			kubeadmutil.CheckErr(err)
 
-			i, err := NewInit(cfgPath, externalcfg, ignorePreflightErrorsSet, skipTokenPrint, dryRun)
+			// TODO: the code in runInit should be progressively converted in phases; each phase will be exposed
+			// via the subcommands automatically created by initRunner.BindToCommand
+			err = runInit(&ctx, out)
 			kubeadmutil.CheckErr(err)
-			kubeadmutil.CheckErr(i.Run(out))
 		},
 	}
 
-	AddInitConfigFlags(cmd.PersistentFlags(), externalcfg, &featureGatesString)
-	AddInitOtherFlags(cmd.PersistentFlags(), &cfgPath, &skipTokenPrint, &dryRun, &ignorePreflightErrors)
-	bto.AddTokenFlag(cmd.PersistentFlags())
-	bto.AddTTLFlag(cmd.PersistentFlags())
+	// adds command flags
+	AddInitConfigFlags(cmd.PersistentFlags(), options.externalcfg, &options.featureGatesString)
+	AddInitOtherFlags(cmd.PersistentFlags(), &options.cfgPath, &options.skipTokenPrint, &options.dryRun, &options.ignorePreflightErrors)
```

Comment:
It will be more clear if this is reduced to `AddInitOtherFlags(cmd.PersistentFlags(), &options)`.

Reply:
Agree with @rosti

Reply:
We can follow up on another PR.